File: samples/AndroidSamples/salmon-android-samples/src/main/java/com/mku/salmon/samples/main/DataStreamActivity.java
package com.mku.salmon.samples.main;

import android.os.Bundle;
import android.widget.Button;
import android.widget.EditText;
import android.widget.Spinner;

import androidx.appcompat.app.AppCompatActivity;

import com.google.android.material.textfield.TextInputEditText;
import com.mku.android.fs.file.AndroidFileSystem;
import com.mku.convert.BitConverter;
import com.mku.salmon.Generator;
import com.mku.salmon.samples.R;
import com.mku.salmon.samples.samples.DataStreamSample;
import com.mku.salmon.samples.samples.SamplesCommon;
import com.mku.salmon.streams.AesStream;
import com.mku.salmon.streams.ProviderType;

import java.util.concurrent.Executor;
import java.util.concurrent.Executors;

public class DataStreamActivity extends AppCompatActivity {
    private TextInputEditText password;
    private Spinner dataSize;
    private EditText outputText;

    private Button encryptButton;
    private Button decryptButton;

    private static final String defaultPassword = "test123";

    private byte[] key;
    private byte[] data;
    private byte[] encData;
    private byte[] nonce;

    private final Executor executor = Executors.newCachedThreadPool();

    @Override
    public void onCreate(Bundle bundle) {
        super.onCreate(bundle);

        AndroidFileSystem.initialize(this);
        AesStream.setAesProviderType(ProviderType.Default);

        setContentView(R.layout.activity_data_stream);

        password = findViewById(R.id.TEXT_PASSWORD);
        dataSize = findViewById(R.id.DATA_SIZE);

        encryptButton = findViewById(R.id.ENCRYPT_BUTTON);
        encryptButton.setOnClickListener((e) -> {
			executor.execute(() -> {
				encryptDataStream();
			});
        });

        decryptButton = findViewById(R.id.DECRYPT_BUTTON);
        decryptButton.setOnClickListener((e) -> {
			executor.execute(() -> {
				decryptDataStream();
			});
        });

        outputText = findViewById(R.id.OUTPUT_TEXT);
        password.setText(defaultPassword);

        initialize();
    }

    public void log(String msg) {
        runOnUiThread(() -> {
            outputText.append(msg + "\n");
        });
    }
	
	public void clearLog() {
        runOnUiThread(() -> {
            outputText.setText("");
        });
    }

    public void initialize() {
        AndroidFileSystem.initialize(this);
        AesStream.setAesProviderType(ProviderType.Default);
    }

    public void encryptDataStream() {
        clearLog();

        // generate a key
        log("generating keys and random data...");
        key = SamplesCommon.getKeyFromPassword(password.getText().toString());

        // Always request a new random secure nonce!
        // if you want to you can embed the nonce in the header data
        // see Encryptor implementation
        nonce = Generator.getSecureRandomBytes(8); // 64 bit nonce
        log("Created nonce: " + BitConverter.toHex(nonce));

        // generate random data
        data = SamplesCommon.generateRandomData(
                Integer.parseInt((String) dataSize.getSelectedItem()) * 1024 * 1024);

        try {
            log("starting encryption...");
            encData = DataStreamSample.encryptDataStream(data, key, nonce, this::log);
        } catch (Exception e) {
            e.printStackTrace();
            log(e.getMessage());
        }
    }

    public void decryptDataStream() {

        try {
            log("starting decryption...");
            byte[] decData = DataStreamSample.decryptDataStream(encData, key, nonce, this::log);
            log("done");
        } catch (Exception e) {
            e.printStackTrace();
            log(e.getMessage());
        }

    }
}

File: samples/DotNetAndroidSamples/DotNetAndroidSamples/Main/DataStreamActivity.cs
using AndroidX.AppCompat.App;
using Android.Widget;
using Android.OS;
using System;
using DotNetAndroidSamples;
using Google.Android.Material.TextField;
using Mku.Android.FS.File;
using Mku.Salmon.Samples.Samples;
using Mku.Salmon.Streams;
using BitConverter = Mku.Convert.BitConverter;
using Android.App;
using System.Threading.Tasks;

namespace Mku.Salmon.Samples.Main;

[Activity(Label = "@string/app_name", Theme = "@style/AppTheme",
    ScreenOrientation = global::Android.Content.PM.ScreenOrientation.Portrait)]
public class DataStreamActivity : AppCompatActivity
{
    private TextInputEditText password;
    private Spinner dataSize;
    private EditText outputText;

    private Button encryptButton;
    private Button decryptButton;

    private static readonly string defaultPassword = "test123";

    private byte[] key;
    private byte[] data;
    private byte[] encData;
    private byte[] nonce;

    protected override void OnCreate(Bundle bundle)
    {
        base.OnCreate(bundle);

        AndroidFileSystem.Initialize(this);
        AesStream.AesProviderType = ProviderType.Default;

        SetContentView(Resource.Layout.activity_data_stream);

        password = (TextInputEditText)FindViewById(Resource.Id.TEXT_PASSWORD);
        dataSize = (Spinner)FindViewById(Resource.Id.DATA_SIZE);

        encryptButton = (Button)FindViewById(Resource.Id.ENCRYPT_BUTTON);
        encryptButton.Click += (s, e) =>
        {
            Task.Run(() =>
            {
                EncryptDataStream();
            });
        };

        decryptButton = (Button)FindViewById(Resource.Id.DECRYPT_BUTTON);
        decryptButton.Click += (s, e) =>
        {
            Task.Run(() =>
            {
                DecryptDataStream();
            });
        };

        outputText = (EditText)FindViewById(Resource.Id.OUTPUT_TEXT);
        password.Text = defaultPassword;

        Initialize();
    }

    public void Log(string msg)
    {
        RunOnUiThread(() =>
        {
            outputText.Append(msg + "\n");
        });
    }

    public void ClearLog()
    {
        RunOnUiThread(() =>
        {
            outputText.Text = "";
        });
    }

    public void Initialize()
    {
        AndroidFileSystem.Initialize(this);
        AesStream.AesProviderType = ProviderType.Default;
    }

    public void EncryptDataStream()
    {
        ClearLog();

        // generate a key
        Log("generating keys and random data...");
        key = SamplesCommon.GetKeyFromPassword(password.Text.ToString());

        // Always request a new random secure nonce!
        // if you want to you can embed the nonce in the header data
        // see Encryptor implementation
        nonce = Generator.GetSecureRandomBytes(8); // 64 bit nonce
        Log("Created nonce: " + BitConverter.ToHex(nonce));

        // generate random data
        data = SamplesCommon.GenerateRandomData(
                int.Parse((string)dataSize.SelectedItem) * 1024 * 1024);

        try
        {
            Log("starting encryption...");
            encData = DataStreamSample.EncryptDataStream(data, key, nonce, Log);
        }
        catch (Exception e)
        {
            Console.Error.WriteLine(e);
            Log(e.Message);
        }
    }

    public void DecryptDataStream()
    {

        try
        {
            Log("starting decryption...");
            byte[] decData = DataStreamSample.DecryptDataStream(encData, key, nonce, Log);
            Log("done");
        }
        catch (Exception e)
        {
            Console.Error.WriteLine(e);
            Log(e.Message);
        }

    }
}
